packages/contracts-periphery/contracts/universal/op-nft/OptimistInviter.sol

```diff
@@ -0,0 +1,228 @@
+
+    /**
+     * @notice Granter who can set accounts' invite counts.
+     */
+    address public immutable INVITE_GRANTER;
+
+    /**
+     * @notice Address of the AttestationStation contract.
+     */
+    AttestationStation public immutable ATTESTATION_STATION;
+
+    /**
+     * @notice Struct that represents a claimable invite that will be signed by the issuer.
+     *
+     * @custom:field issuer   Address that issued the signature. Reason this is explicitly included,
+     *                        and not implicitly assumed to be the recovered address from the
+     *                        signature is that the issuer may be using a ERC-1271 compatible
+     *                        contract wallet, where the recovered address is not the same as the
+     *                        issuer, or the signature is not an ECDSA signature at all.
+     * @custom:field nonce    Pseudorandom nonce to prevent replay attacks.
+     */
+    struct ClaimableInvite {
+        address issuer;
+        bytes32 nonce;
+    }
+
+    /**
+     * @notice Maps from hashes to whether or not they have been committed.
+     */
+    mapping(bytes32 => bool) public commitments;
+
+    /**
+     * @notice Maps from addresses to nonces to whether or not they have been used.
+     */
+    mapping(address => mapping(bytes32 => bool)) public usedNonces;
+
+    /**
+     * @custom:semver 1.0.0
+     *
+     * @param _inviteGranter      Address of the invite granter.
+     * @param _attestationStation Address of the AttestationStation contract.
+     */
+    constructor(address _inviteGranter, AttestationStation _attestationStation) Semver(1, 0, 0) {
+        INVITE_GRANTER = _inviteGranter;
+        ATTESTATION_STATION = _attestationStation;
+    }
+
+    /**
+     * @notice Initializes the OptimistInviter contract, setting the EIP712 context.
+     *
+     * @param _name Contract name
+     */
+    function initialize(string memory _name) public reinitializer(VERSION) {
+        __EIP712_init(_name, version());
+    }
+
+    /**
+     * @notice Allows invite granter to set the number of invites an address has.
+     *
+     * @param _accounts    An array of accounts to update the invite counts of.
+     * @param _inviteCount Number of invites to set to.
+     */
+    function setInviteCounts(address[] calldata _accounts, uint256 _inviteCount) public {
+        // Only invite granter can grant invites
+        require(
+            msg.sender == INVITE_GRANTER,
+            "OptimistInviter: only invite granter can grant invites"
+        );
+
+        uint256 length = _accounts.length;
+
+        for (uint256 i; i < length; ) {
+            // The granted invites are stored as an attestation from this contract on the
+            // AttestationStation contract. Number of invites is stored as a encoded uint256 in the
+            // data field of the attetation.
+            ATTESTATION_STATION.attest(
```

Comment:
Agree on this, then we make a single call instead of `n` calls

Reply:
This uses slightly more gas than the current approach. Was wondering why is making a single call preferable?

```
function setInviteCounts(address[] calldata _accounts, uint256 _inviteCount) public {
        // Only invite granter can grant invites
        require(
            msg.sender == INVITE_GRANTER,
            "OptimistInviter: only invite granter can grant invites"
        );

        uint256 length = _accounts.length;

        AttestationStation.AttestationData[]
            memory attestations = new AttestationStation.AttestationData[](length);

        for (uint256 i; i < length; ) {
            attestations[i] = AttestationStation.AttestationData({
                about: _accounts[i],
                key: bytes32("optimist.can-invite"),
                val: abi.encode(_inviteCount)
            });

            unchecked {
                ++i;
            }
        }

        // The granted invites are stored as an attestation from this contract on the
        // AttestationStation contract. Number of invites is stored as a encoded uint256 in the
        // data field of the attestation.
        ATTESTATION_STATION.attest(attestations);
    }
```

Reply:
Looks like the cost is due to memory expansion while re-encoding. 
I think you could do the original encoding off-chain and just change the arguments to accept an array of attestations.

Reply:
+1 makes ux of using this contract worse but likely worth just dealing with raw bytes

Reply:
I believe you could prevent unecessary allocation of memory by copying the free memory pointer at the top of the loop body and then resetting the free memory pointer at the bottom of the loop body to the prev value. Its a bit of a hack

Reply:
to test gas usage correctly, you need to put the functions in different contracts so that the gas used in function dispatch is controlled for. if there are multiple functions in the same contract the amount of gas to dispatch to them may be different